### PR TITLE
#51: Enabled BFSJ to list buckets on root level

### LINKED ISSUE #51: Listing only unique entries

### PR DESCRIPTION
**Pull request by ckunki**
Fixes #51

### CODE REVIEW

File: src/test/java/com/exasol/bucketfs/ListingProviderIT.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.junit.jupiter.api.Test;
+
+import com.exasol.bucketfs.http.HttpClientBuilder;
+import com.exasol.bucketfs.testutil.BucketCreator;
+
+@Tag("slow")
+class ListingProviderIT extends AbstractBucketIT {
+    private static HttpClient HTTP_CLIENT = new HttpClientBuilder() //
+            // .certificate(null)
+            // .raiseTlsErrors(false)
```
Comment (kaklakariada) [inline src/test/java/com/exasol/bucketfs/ListingProviderIT.java:21]:
```suggestion
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: src/test/java/com/exasol/bucketfs/ClusterConfigurationBucketFactory.java [hunk 2]
```diff
@@ -23,7 +22,7 @@ public final class ClusterConfigurationBucketFactory implements BucketFactory {
      * Create a new instance of a {@link ClusterConfigurationBucketFactory}.
      *
      * @param monitor                      BucketFS synchronization monitor
-     * @param host                    IP address of the the BucketFS service
+     * @param host                         IP address of the the BucketFS service
```
Comment (kaklakariada) [inline src/test/java/com/exasol/bucketfs/ClusterConfigurationBucketFactory.java:25]:
```suggestion
     * @param host                         host name of the the BucketFS service
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/bucketfs/BucketFactory.java [hunk 3]
```diff
@@ -11,5 +11,5 @@ public interface BucketFactory {
      * @param bucketName  name of the bucket
      * @return bucket
      */
-    public Bucket getBucket(final String serviceName, final String bucketName);
+    Bucket getBucket(final String serviceName, final String bucketName);
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/BucketFactory.java:14]: Please don't change this. Our team convention is to explicitly use `public` in interfaces.

File: src/main/java/com/exasol/bucketfs/BucketFsServiceConfigurationProvider.java [hunk 4]
```diff
@@ -5,7 +5,7 @@
 /**
  * Interface for objects that can provide the configuration of a BucketFS service.
  */
-public interface BucketFsSerivceConfigurationProvider {
+public interface BucketFsServiceConfigurationProvider {
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/BucketFsServiceConfigurationProvider.java:8]: Good catch. Once you see it...

File: src/main/java/com/exasol/bucketfs/HttpResponseEvaluator.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package com.exasol.bucketfs;
+
+import static com.exasol.errorreporting.ExaError.messageBuilder;
+import static java.net.HttpURLConnection.*;
+
+import java.net.URI;
+
+/**
+ * THis class provides a central place for evaluation of HTTP response and throws appropriate exceptions.
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/HttpResponseEvaluator.java:9]:
```suggestion
 * This class provides a central place for evaluation of HTTP response and throws appropriate exceptions.
```

File: src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java [hunk 6]
```diff
@@ -38,7 +36,7 @@ public class ReadEnabledBucket implements ReadOnlyBucket {
     /**
      * ip address
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java:37]:
```suggestion
     *Host or IP address
```

File: src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java [hunk 7]
```diff
@@ -57,18 +55,18 @@ public class ReadEnabledBucket implements ReadOnlyBucket {
      * @param builder builder
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java:55]:
```suggestion
     * Create a new instance of a bucket that support reading.
     *
     * @param builder builder from which the bucket should be constructed
```

File: src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java [hunk 8]
```diff
@@ -93,71 +91,18 @@ public List<String> listContents() throws BucketAccessException {
...
-                .map(this::removeLeadingSlash) //
-                .filter(e -> e.startsWith(path)) // keep only entries with path as prefix
-                .map(e -> e.substring(path.length())) // cut of path prefix
-                .map(this::extractFirstPathComponent) //
-                .distinct() // ensure only unique entries
-                .sorted() //
-                .collect(Collectors.toList());
+    public List<String> listContents(final String path) throws BucketAccessException {
+        final String prefix = removeLeadingSeparator(path);
+        return new BucketContentListing(this.client, this.protocol, this.host, this.port, this.bucketName,
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java:99]: Use builder pattern. Too many parameters.

File: src/main/java/com/exasol/bucketfs/list/BucketContentListing.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import com.exasol.bucketfs.BucketAccessException;
+
+/**
+ * This class enables to list the contents of a directory in a bucket.
+ */
+public class BucketContentListing extends ListingProvider {
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/BucketContentListing.java:17]: The name implies a value object, but this class has control behavior.

File: src/main/java/com/exasol/bucketfs/list/BucketContentListing.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+     * Create a new instance
+     *
+     * @param httpClient   HTTP client to use
+     * @param protocol     protocol to use: either "http" or "https"
+     * @param host         host name or IP address
+     * @param port         port of BucketFS service
+     * @param bucketName   name of the bucket to list the contents for
+     * @param readPassword password for reading non-public buckets
+     */
+    public BucketContentListing(final HttpClient httpClient, final String protocol, final String host, final int port,
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/BucketContentListing.java:32]: Add builder, hide constructor.

File: src/main/java/com/exasol/bucketfs/list/BucketContentListing.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+            final String bucketName, final String readPassword) {
+        super(httpClient, protocol, host, port);
+        this.bucketName = bucketName;
+        this.readPassword = readPassword;
+    }
+
+    /**
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/BucketContentListing.java:39]:
```suggestion
    /**
     * Retrieve the listing.
```

File: src/main/java/com/exasol/bucketfs/list/BucketListing.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.exasol.bucketfs.BucketAccessException;
+
+/**
+ * This class enables to retrieve a list of buckets.
+ */
+public class BucketListing extends ListingProvider {
+
+    /**
+     * Create a new instance
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/BucketListing.java:18]: Class name link missing. Trailing period missing.

File: src/main/java/com/exasol/bucketfs/list/ListingProvider.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.Base64;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+
+import com.exasol.bucketfs.*;
+
+/**
+ * This class enables to request list of buckets or of objects inside a bucket.
+ */
+public abstract class ListingProvider {
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/ListingProvider.java:20]:
```suggestion
public abstract class AbstractListingProvider {
```

Rename other instance accordingly, please.

File: src/main/java/com/exasol/bucketfs/list/ListingProvider.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * Retrieve the content for the given URI as stream.
+     *
+     * @param uri          URI to retrieve listing for
+     * @param readPassword read password, only relevant for non-public buckets
+     * @return stream of strings
+     * @throws BucketAccessException in case of failure
+     */
+    protected Stream<String> listingStream(final URI uri, final String readPassword) throws BucketAccessException {
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/ListingProvider.java:59]: Start function name with a verb.

Recommened:

`listUriContentAsStream`

File: src/main/java/com/exasol/bucketfs/list/ListingProvider.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        return Arrays.stream(requestListing(uri, readPassword).split("\\s+")).sorted();
+    }
+
+    private String requestListing(final URI uri, final String readPassword) throws BucketAccessException {
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/ListingProvider.java:63]: Where is the version for public buckets without password?

File: src/test/java/com/exasol/bucketfs/ListingProviderIT.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+import org.junit.jupiter.api.Test;
+
+import com.exasol.bucketfs.http.HttpClientBuilder;
+import com.exasol.bucketfs.list.BucketContentListing;
+import com.exasol.bucketfs.list.BucketListing;
+import com.exasol.bucketfs.testutil.BucketCreator;
+
+@Tag("slow")
+class ListingProviderIT extends AbstractBucketIT {
+
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/ListingProviderIT.java:20]:
```suggestion
class ListingProviderIT extends AbstractBucketIT {
```

File: src/test/java/com/exasol/bucketfs/ListingProviderIT.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    private static HttpClient HTTP_CLIENT = new HttpClientBuilder().build();
+
+    @Test
+    void testListBuckets() throws Exception {
+        final BucketCreator bucketCreator = bucketCreator().assumeJsonRpcAvailable().createBucket();
+
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/ListingProviderIT.java:26]:
```suggestion
        final BucketCreator bucketCreator = bucketCreator().assumeJsonRpcAvailable().createBucket();
```

File: src/test/java/com/exasol/bucketfs/ListingProviderIT.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+        bucketCreator.waitUntilBucketExists();
+        final List<String> actual = listBuckets();
+        assertThat(actual, hasItems(bucketCreator.getBucketName(), "default"));
+    }
+
+    @Test
+    void listBucketContentsWithoutReadPassword() throws Exception {
+        final BucketCreator bucketCreator = bucketCreator().assumeJsonRpcAvailable().createBucket();
+
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/ListingProviderIT.java:35]:
```suggestion
        final BucketCreator bucketCreator = bucketCreator().assumeJsonRpcAvailable().createBucket();
```

File: src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java [hunk 19]
```diff
@@ -42,28 +31,48 @@ void testGetDefaultBucket() {
 
     // [itest->dsn~bucket-lists-its-contents~2]
     @Test
-    void testListBucketContentsWithRootPath() throws BucketAccessException {
+    void listBucketContentsWithRootPath() throws BucketAccessException {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java:34]: Start all test functions with `test` please.

File: src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java [hunk 20]
```diff
@@ -42,28 +31,48 @@ void testGetDefaultBucket() {
...
         assertThat(getDefaultBucket().listContents(pathInBucket), hasItem(startsWith("ScriptLanguages")));
     }
 
     @Test
-    void testListBucketContentsOfIllegalPathThrowsException() {
+    void listContentsWithWrongReadPassword_Fails() throws Exception {
+        final String bucketName = createBucket("dir/file.txt", "protected content");
+        final ReadOnlyBucket bucket = getBucket(bucketName, "wrong read password");
+        ExceptionAssertions.assertThrowsWithMessage(BucketAccessException.class, () -> bucket.listContents("dir/"), //
+                startsWith("E-BFSJ-3: Access denied trying to list "));
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java:50]: :+1:

File: src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java [hunk 21]
```diff
@@ -42,28 +31,48 @@ void testGetDefaultBucket() {
...
+        final String pathInBucket = "dir/file.txt";
+        final String bucketName = createBucket(pathInBucket, "protected content");
+        final ReadOnlyBucket bucket = getBucket(bucketName, "wrong read password");
+        ExceptionAssertions.assertThrowsWithMessage(BucketAccessException.class,
+                () -> bucket.downloadFileAsString(pathInBucket), //
+                startsWith("E-BFSJ-3: Access denied trying to download "));
+    }
+
+    @Test
+    void listBucketContentsOfIllegalPath_ThrowsException() {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java:64]: No Underscores in method names please.

```suggestion
    void testListingBucketContentsOfIllegalPathThrowsException() {
```

Comment (redcatbear) [reply src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java:64]: Similar in other tests.

File: src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java [hunk 22]
```diff
@@ -74,17 +83,50 @@ void testDownloadFileThrowsExceptionOnIllegalPathInBucket(@TempDir final Path te
...
+                .host(getHost()) //
                 .httpPort(getMappedDefaultBucketFsPort()) //
                 .serviceName(DEFAULT_BUCKETFS) //
                 .name(DEFAULT_BUCKET) //
                 .readPassword(getDefaultBucketConfiguration().getReadPassword()) //
                 .build();
         assertThat(bucket.listContents(), hasItem("EXAClusterOS/"));
     }
+
+    // ------------------------------------------------------
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java:100]:
```suggestion
```

File: src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java [hunk 23]
```diff
@@ -74,17 +83,50 @@ void testDownloadFileThrowsExceptionOnIllegalPathInBucket(@TempDir final Path te
...
+
+    private ReadOnlyBucket getDefaultBucket() {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java:102]: Order methods by first use.

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/bucketfs/HttpResponseEvaluator.java [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+ * This class provides a central place for evaluation of HTTP response and throws appropriate exceptions.
+ */
+public class HttpResponseEvaluator {
+
+    private HttpResponseEvaluator() {
+        // only static usage
+    }
+
+    /**
+     * Evaluates the response
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/HttpResponseEvaluator.java:18]:
```suggestion
     * Evaluates the HTTP response code.
     * <p>
     * In case of success does nothing, otherwise throws appropriate exception.
     * </p>
```

File: src/main/java/com/exasol/bucketfs/list/BucketContentLister.java [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.stream.Collectors;
+
+import com.exasol.bucketfs.BucketAccessException;
+
+/**
+ * This class enables to list the contents of a directory in a bucket.
+ */
+public class BucketContentLister {
+
+    private final URI bucketUri;
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/BucketContentLister.java:19]:
```suggestion
public class BucketContentLister {
    private final URI bucketUri;
```

File: src/main/java/com/exasol/bucketfs/list/BucketService.java [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.stream.Collectors;
+
+import com.exasol.bucketfs.BucketAccessException;
+
+/**
+ * This class enables to retrieve a list of buckets.
+ */
+public class BucketService {
+
+    private final ListingRetriever listingRetriever;
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/BucketService.java:16]:
```suggestion
public class BucketService {
    private final ListingRetriever listingRetriever;
```

File: src/main/java/com/exasol/bucketfs/list/BucketService.java [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+                .retrieve(this.bucketServiceUri, "") //
+                .collect(Collectors.toList());
+        if (list.isEmpty()) {
+            throw pathToBeListedNotFoundException();
+        } else {
+            return list;
+        }
+    }
+
+    private BucketAccessException pathToBeListedNotFoundException() {
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/BucketService.java:45]: Extracting this code seems unnecessary to me. Also: the function name does not start with a verb. :-)

File: src/main/java/com/exasol/bucketfs/list/ListingRetriever.java [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.stream.Stream;
+
+import com.exasol.bucketfs.*;
+
+/**
+ * This class retrieves the list of buckets or of objects inside a bucket.
+ */
+public class ListingRetriever {
+
+    /**
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/list/ListingRetriever.java:22]:
```suggestion
public class ListingRetriever {
    /**
```

File: src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java [hunk 29]
```diff
@@ -64,62 +43,80 @@ private Path createTestFile(final Path tempDir, final String fileName, final int
...
-            throws BucketAccessException, InterruptedException, IOException, TimeoutException {
+    @ValueSource(strings = { "dir1/", "dir2/sub2/", "dir3/sub3/subsub3/", "/dir4/", "/dir5/sub5/" })
+    void testUploadToDirectoryInBucket(final String pathInBucket, @TempDir final Path tempDir) throws Exception {
         final var fileName = "file.txt";
         final var file = createTestFile(tempDir, fileName, 1);
-        final var bucket = createDefaultBucket();
+        final var bucket = getDefaultBucketForWriting();
         bucket.uploadFile(file, pathInBucket);
-        assertThat(createDefaultBucket().listContents(pathInBucket), Matchers.contains(fileName));
+        assertThat(getDefaultBucketForWriting().listContents(pathInBucket), Matchers.contains(fileName));
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java:53]: What this does not test is what happens if the path contains more than one file. That listing could still be wrong.

File: src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java [hunk 30]
```diff
@@ -64,62 +43,80 @@ private Path createTestFile(final Path tempDir, final String fileName, final int
...
     }
 
     // [itest->dsn~uploading-strings-to-bucket~1]
     @Test
-    void testUploadStringContent() throws IOException, BucketAccessException, InterruptedException, TimeoutException {
+    void testUploadStringContent() throws Exception {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java:58]: Are you sure, Sonar is OK with this contraction of exceptions? Because I am skeptical. Sure, this is more compact, but it also hides what are "valid" exceptions that can be thrown here.

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 31]
```diff
@@ -4,70 +4,46 @@
...
-
-        final CreateBucketCommandBuilder command = createCommandFactory(true).makeCreateBucketCommand() //
-                .bucketFsName(DEFAULT_BUCKETFS) //
-                .bucketName(randomBucketName) //
-                .isPublic(true) //
-                .readPassword(READ_PASSWORD) //
-                .writePassword(WRITE_PASSWORD);
-
-        final JsonRpcException exception = assertThrows(JsonRpcException.class, () -> command.execute());
+        final BucketCreator bucketCreator = bucketCreator().assumeJsonRpcAvailable();
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:22]: This refactoring is a bit out of scope of the current ticket. Next time I suggest to split it off into its own PR. That would also make reviewing easier.

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 32]
```diff
@@ -76,125 +52,37 @@ void testCreatingBucketWithCertificateFailsHostnameValidation() throws BucketAcc
...
     @Test
-    void testCreatedBucketIsWriteable() throws BucketAccessException, TimeoutException, InterruptedException {
-        assumeJsonRpcAvailable();
+    void testCreatedBucketWithDefaultValues() throws BucketAccessException, TimeoutException, InterruptedException {
+        final BucketCreator bucketCreator = bucketCreator().assumeJsonRpcAvailable();
+        assertDoesNotThrow(() -> bucketCreator.commandWithDefaultValues().execute());
+    }
 
-        final String randomBucketName = getUniqueBucketName();
+    // -----------------------------------------------------
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:77]:
```suggestion
```
No ASCII art please.

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java (hunk 29 above)
Comment (ckunki) [reply src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java:53]: See [doc](http://hamcrest.org/JavaHamcrest/javadoc/1.3/org/hamcrest/Matchers.html#contains(E...)): 
> For a positive match, the examined iterable must be of the same length as the number of specified items.

File: src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java (hunk 30 above)
Comment (ckunki) [reply src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java:58]: Rationale for my changes 
* increased readability, less clutter, less code 
* no loss in functionality, expressiveness, no increased risk for unidentified bugs as test is intended to not throw an exception anyway

**Review by redcatbear — APPROVED** [review]

**Review by redcatbear — APPROVED** [review]
